packages/eslint-plugin/tests/rules/strict-void-return.test.ts

```diff
@@ -0,0 +1,2767 @@
+      errors: [
+        {
+          column: 45,
+          data: { varName: 'cb' },
+          line: 1,
+          messageId: 'nonVoidReturnInVar',
+        },
+      ],
+    },
+    {
+      code: `
+        const cb: () => void = (): Array<number> => {
+          return [];
+        };
+      `,
+      errors: [
+        {
+          column: 36,
+          data: { varName: 'cb' },
+          line: 2,
+          messageId: 'nonVoidFuncInVar',
+        },
+      ],
+    },
+    {
+      code: noFormat`const cb: () => void = function*foo() {}`,
+      errors: [
+        {
+          column: 24,
+          data: { varName: 'cb' },
+          line: 1,
+          messageId: 'nonVoidFuncInVar',
+        },
+      ],
+    },
+    {
+      code: 'const cb: () => void = (): Promise<number> => Promise.resolve(1);',
+      errors: [
+        {
+          column: 47,
+          data: { varName: 'cb' },
+          line: 1,
+          messageId: 'nonVoidReturnInVar',
+        },
+      ],
+      options: [{ allowReturnUndefined: false }],
+    },
+    {
+      code: `
+        const cb: () => void = async (): Promise<number> => {
+          try {
+            return Promise.resolve(1);
+          } catch {}
+        };
+      `,
+      errors: [
+        {
+          column: 42,
+          data: { varName: 'cb' },
+          line: 2,
+          messageId: 'nonVoidFuncInVar',
+        },
+      ],
+      options: [{ allowReturnUndefined: false }],
+    },
+    {
+      code: 'const cb: () => void = async (): Promise<number> => Promise.resolve(1);',
+      errors: [
+        {
+          column: 50,
+          data: { varName: 'cb' },
+          line: 1,
+          messageId: 'asyncFuncInVar',
+        },
+      ],
+    },
+    {
+      code: `
+        const foo: () => void = async () => {
+          try {
+            return 1;
+          } catch {}
+        };
+      `,
+      errors: [
+        {
+          column: 13,
+          data: { varName: 'foo' },
+          line: 4,
+          messageId: 'nonVoidReturnInVar',
+        },
+      ],
+    },
+    {
+      code: `
+        const foo: () => void = async (): Promise<void> => {
+          try {
+            await Promise.resolve();
+          } finally {
+          }
+        };
+      `,
+      errors: [
+        {
+          column: 57,
+          data: { varName: 'foo' },
+          line: 2,
+          messageId: 'asyncFuncInVar',
+        },
+      ],
+    },
+    {
+      code: `
+        const foo: () => void = async () => {
+          try {
+            await Promise.resolve();
+          } catch (err) {
+            console.error(err);
+          }
+          console.log('ok');
+        };
+      `,
+      errors: [
+        {
+          column: 42,
+          data: { varName: 'foo' },
+          line: 2,
+          messageId: 'asyncFuncInVar',
+        },
+      ],
+      options: [{ allowReturnUndefined: false }],
+    },
+    // TODO: Check every union type separately
```

Comment:
[Question] Is this still pending? What's up?

Reply:
This one was non-trivial to implement and I think it's not worth it.